File: quicksort.py
#quicksort algorithm implemented recursively
arrz = [78,45,39,938,3,44,64,47]
def quicksorrt(arrz):
	if len(arrz) <=1:
		return arrz
	pivot = arrz[len(arrz)//2]
	lesser = []
	equal = []
	greater = []

	for y in range(len(arrz)):
		if arrz[y] < pivot:
			lesser.append(arrz[y])
		elif arrz[y] == pivot:
			equal.append(arrz[y])
		else:
			greater.append(arrz[y])
	return quicksorrt(lesser) + equal + quicksorrt(greater)

print("Quiksorted array/list:", quicksorrt(arrz))
print()






#quicksort algorithm implemented recursively in python
array = [2,7,8,17,98,4]
def qsort(array):
	if len(array) <=1:
		return array
	pivot = array[len(array)//2]
	lessr = []
	equal = []
	greater = []

	for z in range(len(array)):
		if array[z] < pivot:
			lessr.append(array[z])
		elif array[z] == pivot:
			equal.append(array[z])
		else:
			greater.append(array[z])

	return qsort(lessr) + equal + qsort(greater)

print("QUICKSORTED ARRAY:", qsort(array))
print()







#QUICKSORT ALGORITHYM implemented using recursion
arr = [6,3,7,9,2,0,1]
def qsort(arr):
    if len(arr) <= 1:
        return arr
    else:
        pivot = arr[len(arr)//2]
        lesser = []
        equal = []
        greater = []
        
        for i in range(len(arr)):
            if(arr[i] < pivot):
                lesser.append(arr[i])
            elif arr[i] == pivot:
                equal.append(arr[i])
            else:
                greater.append(arr[i])
        
        return qsort(lesser) + equal + qsort(greater)

print("Quicksorted list:", qsort(arr))



arr = [5,545,8,4,97,2132,2,1,9]

def qsort(arr):
    if len(arr) <= 1:
        return arr
    else:
        pivot = arr[len(arr)//2]
        lesser = []
        greater = []

        for i in range(len(arr)):
            if arr[i] < pivot:
                lesser.append(arr[i])
            elif arr[i] > pivot:
                greater.append(arr[i])
        return qsort(lesser) + [pivot] + qsort(greater)

print("The quicksorted array:", qsort(arr))


arr = [6,9,3,0,24,67,2,1234]

def quicksort(arr):
    if len(arr) <= 1:
        return arr
    
    pivot = arr[len(arr)//2]
    lesser = []
    greater = []

    for i in range(len(arr)):
        if arr[i] < pivot:
            lesser.append(arr[i])
        elif arr[i] > pivot:
            greater.append(arr[i])
    return quicksort(lesser) + [pivot] + quicksort(greater)

print("Quicksorted list/array:", quicksort(arr))


#Quicksort algorithm implemented recursively...choosing middle element as pivot
def quick_sort(arr):
	if len(arr) <=1:
		return arr
	else:
		pivot = arr[len(arr)//2]
		lesser = []
		greater = [] 

		for i in range(len(arr)):
			if arr[i] < pivot:
				lesser.append(arr[i])
			elif arr[i] > pivot:
				greater.append(arr[i])
		return quick_sort(lesser) + [pivot] + quick_sort(greater)
print("Quicksorted:", quick_sort([5,2,3,1,4,0]))


#Quicksort algorithm implemented recursively...an implementation involving choosing the pivot randomly
import random

def quicksort(arr):
	if len(arr) <= 1:
		return arr 
	else:
		pivot = random.choice(arr)
		lesser = []
		greater = []

		for elem in range(len(arr)):
			if arr[elem] < pivot:
				lesser.append(arr[elem])
			elif arr[elem] > pivot:
				greater.append(arr[elem])

		return quicksort(lesser) + [pivot] + quicksort(greater)

print("sorted list:", quicksort([5,3,9,0,2,4,1,6,8,7]))


#quicksort algorithm in python using recursion...mastery 1
arr = [6,5,4,3,2,1]

def quicksort(arr):
    if len(arr) <= 1:
        return arr
    #Middle index as pivot...value floored
    pivot = arr[len(arr)//2]
    lessr = []
    greatr = []
    
    for i in range(len(arr)):
        if arr[i] < pivot:
            lessr.append(arr[i])
        elif arr[i] > pivot:
            greatr.append(arr[i])
    return quicksort(lessr) + [pivot] + quicksort(greatr)

print("Quicksorted list:", quicksort(arr))


#Another quicksort implemented recursively
arry = [5,2,1,3,4]
def quick_sort(arry):
	if len(arry) <= 1:
		return arry
	pivot = arry[len(arry)//2]
	lesser = []
	greater = []
	
	for i in range(len(arry)):
		if arry[i] < pivot:
			lesser.append(arry[i])
		elif arry[i] > pivot:
			greater.append(arry[i])
	return quick_sort(lesser) + [pivot] + quick_sort(greater)

print("Quicksorted list:", quick_sort(arry))




#Quicksort implemented recursively...takes integers from user and uses them to populate list arry
arry = []
for i in range(5):
	inputt = int(input("Enter a number:"))
	arry.append(inputt)
	
def quick_sort(arry):
	if len(arry) <= 1:
		return arry
	pivot = arry[len(arry)//2]
	lesser = []
	greater = []
	
	for i in range(len(arry)):
		if arry[i] < pivot:
			lesser.append(arry[i])
		elif arry[i] > pivot:
			greater.append(arry[i])
	return quick_sort(lesser) + [pivot] + quick_sort(greater)

print("Quicksorted list:", quick_sort(arry))



#Quicksort algorithm implemented recursively.
#It takes user input and uses the integers to populate a list
array = []

for x in range(5):
	y = int(input("Enter a number:"))
	array.append(y)

def quicksort(array):
	if len(array) <= 1:
		return array

	pivot = array[len(array)//2]
	lesser = []
	greater = []

	for index in range(len(array)):
		if array[index] < pivot:
			lesser.append(array[index])
		elif array[index] > pivot:
			greater.append(array[index])
	return quicksort(lesser) + [pivot] + quicksort(greater)

print("Quicksorted array:", quicksort(array))



#Quicksort...implemented recursively
arr = [6,5,2,9,1]

def quick_sort(arr):
	if len(arr) <= 1:
		return arr

	pivot = arr[len(arr) //2 ]
	lesser= []
	greater = []

	for elem in range(len(arr)):
		if arr[elem] < pivot:
			lesser.append(arr[elem])
		elif arr[elem] > pivot:
			greater.append(arr[elem])
	return quick_sort(lesser) + [pivot] + quick_sort(greater)

print("Quicksorted list:", quick_sort(arr))


#Quicksort algorithm implemented recursively
arry = [6,3,98,3903,4,2,9,4,1,0]

def quicksort(arry):
	if len(arry) <= 1:
		return arry
	pivot = arry[len(arry) // 2]
	lesser = []
	greater = []

	for elem in range(len(arry)):
		if arry[elem] < pivot:
			lesser.append(arry[elem])
		elif arry[elem] > pivot:
			greater.append(arry[elem])
			#skips over arry[elem] == pivot
	return quicksort(lesser) + quicksort([pivot]) + quicksort(greater)

print("Quicksorted array/list:", quicksort(arry))


#Another quicksort algorithm example implemented recursively
arr = [65,3,9,0,43,1,7,0]

def quicksorto(arr):
	if len(arr) <= 1:
		return arr
	pivot = arr[len(arr) // 2]
	lesser = []
	greater = []

	for item in range(len(arr)):
		if arr[item] < pivot:
			lesser.append(arr[item])
		elif arr[item] > pivot:
			greater.append(arr[item])
	return quicksorto(lesser) + [pivot] + quicksorto(greater)

print("Quicksorted array:", quicksorto(arr))



#Quicksort algorithm...recursively
arr = [4,2,1,3,0]

def quicksort(arr):
	if len(arr) <= 1:
		return arr 

	pivot = arr[len(arr)//2]
	lesser = []
	greater = []

	for i in range(len(arr)):
		if arr[i] < pivot:
			lesser.append(arr[i])
		elif arr[i] > pivot:
			greater.append(arr[i])
	return quicksort(lesser) + [pivot] + quicksort(greater)

print("The Quicksorted list:", quicksort(arr))



arr=[78,4,8,2,7,9,2334]

def q_sort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr)//2]
    lessr = []
    greatr = []
   
    for i in range(len(arr)):
        if arr[i] < pivot:
            lessr.append(arr[i])
        elif arr[i] > pivot:
            greatr.append(arr[i])
    return q_sort(lessr) + [pivot] + q_sort(greatr)

print("Quicksorted array:", q_sort(arr))



arr4y = [8,9,1,4,7,0,2,6]

def quickssort(arr4y):
	if len(arr4y) <= 1:
		return arr4y
	pivot = arr4y[len(arr4y)//2]
	lessr = []
	greatr = []

	for item in range(len(arr4y)):
		if arr4y[item] < pivot:
			lessr.append(arr4y[item])
		elif arr4y[item] > pivot:
			greatr.append(arr4y[item])
	return quickssort(lessr) + [pivot] + quickssort(greatr)
print("Array after quicksort:", quickssort(arr4y))


#QUICKSORT ALGORITHYM
arr = [6,9,3,0,24,67,2,1234]

def quicksort(arr):
    if len(arr) <= 1:
        return arr
    
    pivot = arr[len(arr)//2]
    lesser = []
    greater = []

    for i in range(len(arr)):
        if arr[i] < pivot:
            lesser.append(arr[i])
        elif arr[i] > pivot:
            greater.append(arr[i])
    return quicksort(lesser) + [pivot] + quicksort(greater)

print("Quicksorted list/array:", quicksort(arr))
            
    

           



	

File: quicksort.js
//Quicksort algorithm recursive implementation
	function quicksort(arr){
		if(arr.length <= 1){
			return arr;
		}
		else{
			const pivot = arr[Math.floor(arr.length/2)];
			const lesser = [];
			const greater = [];

			for(let m = 0; m < arr.length; m++){
				if(arr[m] < pivot){
					lesser.push(arr[m]);
				}
				else if(arr[m] > pivot){
					greater.push(arr[m]);
				}
			}
			return quicksort(lesser).concat(pivot, quicksort(greater));
		}
	}
	console.log("Quicksort results: " + quicksort([3,4,1,7,0,2,5]));

	
	//Quicksort algorithm in ES6 recursively...using Arrow functions
	const quicksrrt = array => { 
		if(array.length <= 1){
			return array;
		}
		else{
			const pivot = array[Math.floor(array.length/2)];
			const left =[];
			const right =[];

			for(i=0; i<array.length; i++){
				if(array[i] < pivot){
					left.push(array[i]);
				}
				else if(array[i]>pivot){
					right.push(array[i]);
				}
			}
			return quicksrrt(left).concat(pivot, quicksrrt(right));
		}
	}
	console.log("result:" + quicksrrt([3,1,2,4,0]));



//quicksort algorithm in ES6 using recursion...mastery 1
arr = [4,3,2,1];

const quicksort = (arr) => {
    if(arr.length <= 1){
        return arr;
    }
    //choosing middle index
    const pivot =arr[Math.floor(arr.length/2)];
    const lesser = [];
    const greater = [];
    
    for(let x=0;x<arr.length;x++){
        if(arr[x] < pivot){
            lesser.push(arr[x]);
        }
        else if(arr[x] > pivot){
            greater.push(arr[x]);
        }
    }
    return quicksort(lesser).concat([pivot], quicksort(greater));
}

console.log("Quicksorted array:", quicksort(arr))



//Another quicksort algorithm implemented recursively
const arr = [5,686,6,90,7,1];

const quicksorto = (arr) => {
	if (arr.length <= 1){
		return arr;
	}
	else{
		const pivot = arr[Math.floor(arr.length / 2)];
		const lesser = [];
		const greater = [];

		for(let i=0; i<arr.length; i++){
			if(arr[i] < pivot){
				lesser.push(arr[i]);
			}
			else if(arr[i] > pivot){
				greater.push(arr[i]);
			}
		}

		return quicksorto(lesser).concat([pivot], quicksorto(greater));

	}
}

console.log("Quickosorted array:", quicksorto(arr))




//Quicksort algorithm implemented recursively
let arry = [7,2,5,1,0,4,3];

const quick_sort = (arry) => {
	if(arry.length <= 1){
		return arry;
	}
	else{
		let pivot = arry[Math.floor(arry.length/2)];
		let lesser = [];
		let greater = [];

		for(let i=0; i<arry.length; i++){
			if(arry[i] < pivot){
				lesser.push(arry[i]);
			}
			else if(arry[i] > pivot){
				greater.push(arry[i]);
			}
		}
		return quick_sort(lesser).concat([pivot], quick_sort(greater));
	}
}

console.log("The quicksorted list is:", quick_sort(arry));



//Quicksort algorithm in es6 javascript
const arr_ = [4,98,574,0,2,7,1,5,3];

const quicksorrrt = (arr_) => {
	if(arr_.length <=1){
		return arr_;
	}
	else{
		let pivot = arr_[Math.floor(arr_.length/2)];
		let lessr = [];
		let greatr = [];

		for(let i=0; i<= arr_.length;i++){
			if(arr_[i] < pivot){
				lessr.push(arr_[i]);
			}
			else if(arr_[i] > pivot){
				greatr.push(arr_[i]);
			}
		}
		return quicksorrrt(lessr).concat([pivot], quicksorrrt(greatr));
	}
}
process.stdout.write(`Old unsorted array was: ${JSON.stringify(arr_)}`);
console.log();
console.log("The new Quicksorted array:", quicksorrrt(arr_));






